### PR TITLE
PR for assessment submission

### PR DESCRIPTION
**Pull request by benjmm**

### CODE REVIEW

File: echo.py [hunk 1]
```diff
@@ -2,21 +2,48 @@
...
 """An enhanced version of the 'echo' cmd line utility"""
 
-__author__ = "???"
+__author__ = "benjmm"
 
 
 import sys
+import argparse
+if sys.version_info[0] >= 3:
+    raise Exception("This program requires python2 interpreter")
```
Comment (Janell-Huyck) [inline echo.py]: :+1:

File: echo.py [hunk 2]
```diff
@@ -2,21 +2,48 @@
...
+        '-t', '--title', help='convert text to titlecase', action='store_true')
+    parser.add_argument('text', help='text to be manipulated')
+    return parser
 
 
 def main(args):
     """Implementation of echo"""
-    pass
+    parser = create_parser()
+    ns = parser.parse_args(args)
```
Comment (Janell-Huyck) [inline echo.py]: what's "ns" mean/stand for?  to my eyes it looks like random letters.

File: tests/test_echo.py [hunk 3]
```diff
@@ -3,10 +3,68 @@
...
+    def test_multiple_2(self):
+        """-ul options are stored and heLLo becomes hello"""
+        args = ["heLLo!", "-ul"]
+        namespace = self.parser.parse_args(args)
+        self.assertTrue(namespace.upper)
+        self.assertTrue(namespace.lower)
+        self.assertEquals(echo.main(args), "hello!")
+
+    def test_noargs(self):
+        """No arguments given and unaltered input text is returned"""
```
Comment (Janell-Huyck) [inline tests/test_echo.py]: i'm thinking that the "text" is an argument here, and would maybe reword the help for this test.

File: tests/test_echo.py [hunk 4]
```diff
@@ -3,10 +3,68 @@
...
+        self.assertTrue(namespace.lower)
+        self.assertTrue(namespace.title)
+        self.assertEquals(echo.main(args), "Hello!")
+
+    def test_multiple_2(self):
+        """-ul options are stored and heLLo becomes hello"""
+        args = ["heLLo!", "-ul"]
+        namespace = self.parser.parse_args(args)
+        self.assertTrue(namespace.upper)
+        self.assertTrue(namespace.lower)
```
Comment (Janell-Huyck) [inline tests/test_echo.py]: this would be a great place to throw in a
`self.assertFalse(namespace.title)`

:)

File: tests/test_echo.py [hunk 5]
```diff
@@ -3,10 +3,68 @@
...
+        """Running the program without arguments shows usage"""
+        # Run the command `python ./echo.py -h` in a separate process, then
+        # collect its output.
+        process = subprocess.Popen(
+            ["python", "./echo.py", "-h"],
+            stdout=subprocess.PIPE)
+        stdout, _ = process.communicate()
+        usage = open("./USAGE", "r").read()
+
+        self.assertEquals(stdout, usage)
```
Comment (Janell-Huyck) [inline tests/test_echo.py]: this test as written tests if your help message is the same as the usage file.  your comment on line 14 implies you're testing the program without any arguments, which would be `["python", "./echo.py"]`.  Just a different test.

it would be simple enough to create another test that tests if `["python", "./echo.py"]` is the same thing as usage as well.

File: tests/test_echo.py [hunk 6]
```diff
@@ -3,10 +3,68 @@
```
Comment (Janell-Huyck) [inline tests/test_echo.py]: overall this looks really good!  you have nice, clean tests that are easy to understand.  the test names are consistent and appropriate.  you're testing almost all the conditions asked for.  see note under line 23 for the one test i think you're missing.

File: echo.py (hunk 2 above)
Comment (benjmm) [reply echo.py]: namespace

File: tests/test_echo.py (hunk 4 above)
Comment (benjmm) [reply tests/test_echo.py]: Good point; didn't think of that. Added.

File: tests/test_echo.py (hunk 5 above)
Comment (benjmm) [reply tests/test_echo.py]: How about I just change the comment? It was included with the starter code. :P

File: tests/test_echo.py (hunk 3 above)
Comment (benjmm) [reply tests/test_echo.py]: Changed to """No optional arguments given and unaltered input text is returned"""
